cli/tools/bundle/mod.rs

```diff
@@ -0,0 +1,868 @@
+    | protocol::ImportKind::ComposesFrom
+    | protocol::ImportKind::DynamicImport
+    | protocol::ImportKind::ImportRule
+    | protocol::ImportKind::UrlToken => ResolutionMode::Import,
+    protocol::ImportKind::RequireCall
+    | protocol::ImportKind::RequireResolve => ResolutionMode::Require,
+  }
+}
+
+impl DenoPluginHandler {
+  #[allow(dead_code)]
+  fn get_final_path(
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<String, AnyError> {
+    match specifier.scheme() {
+      "file" => {
+        let path = specifier.to_file_path().unwrap();
+        Ok(path.to_string_lossy().to_string())
+      }
+      "npm" => {
+        let req_ref = NpmPackageReqReference::from_specifier(specifier)?;
+        let path = self.npm_req_resolver.resolve_req_reference(
+          &req_ref,
+          specifier,
+          ResolutionMode::Import,
+          NodeResolutionKind::Execution,
+        )?;
+        Ok(path.to_string_lossy().to_string())
+      }
+      "https" | "http" => Ok(match &self.http_cache {
+        GlobalOrLocalHttpCache::Global(global_http_cache) => global_http_cache
+          .local_path_for_url(specifier)?
+          .to_string_lossy()
+          .to_string(),
+        GlobalOrLocalHttpCache::Local(local_http_cache) => local_http_cache
+          .local_path_for_url(specifier)?
+          .ok_or(BundleError::HttpCache)?
+          .to_string_lossy()
+          .to_string(),
+      }),
+      _ => Ok(specifier.to_string()),
+    }
+  }
+  fn bundle_resolve(
+    &self,
+    path: &str,
+    importer: Option<&str>,
+    resolve_dir: Option<&str>,
+    kind: esbuild_rs::protocol::ImportKind,
+    // TODO: use this / store it for later usage when loading
+    with: IndexMap<String, String>,
+  ) -> Result<Option<String>, AnyError> {
+    log::debug!(
+      "bundle_resolve: {:?} {:?} {:?} {:?} {:?}",
+      path,
+      importer,
+      resolve_dir,
+      kind,
+      with
+    );
+    let mut resolve_dir = resolve_dir.unwrap_or("").to_string();
+    let resolver = self.resolver.clone();
+    if !resolve_dir.ends_with(std::path::MAIN_SEPARATOR) {
+      resolve_dir.push(std::path::MAIN_SEPARATOR);
+    }
+    let resolve_dir_path = Path::new(&resolve_dir);
+    let mut referrer =
+      resolve_url_or_path(importer.unwrap_or(""), resolve_dir_path)
+        .unwrap_or_else(|_| {
+          Url::from_directory_path(std::env::current_dir().unwrap()).unwrap()
+        });
+    if referrer.scheme() == "file" {
+      let pth = referrer.to_file_path().unwrap();
+      if (pth.is_dir()) && !pth.ends_with(std::path::MAIN_SEPARATOR_STR) {
+        referrer.set_path(&format!(
+          "{}{}",
+          referrer.path(),
+          std::path::MAIN_SEPARATOR
+        ));
+      }
+    }
+
+    log::debug!(
+      "{}: {} {} {} {:?}",
+      deno_terminal::colors::magenta("op_bundle_resolve"),
+      path,
+      resolve_dir,
+      referrer,
+      import_kind_to_resolution_mode(kind)
+    );
+    let result = resolver.resolve(
```

Comment:
I added a `resolve_with_graph` method to `DenoResolver` a few days ago. I think we can use that here? It should help reduce some of this logic.